README
AUTOTESTING README
==================

autotesting --tests=TESTS.XML

This readme outlines the different tags used in TESTS.XML. 

EXAMPLE TESTS.XML
=================

<autotesting>
  <tests>
    <test> 
      <download>
        <url>http://127.0.0.1/autotesting/debian-live-501-i386-xfce-desktop.iso</url>
      </download>
      <title>Debian Live (xfce)</title>
      <description>Debian Live. Daily build of squeeze xfce iso.</description>
      <background>http://git.debian.org/?p=debian-live/homepage.git;a=blob_plain;f=images/debian-live.png</background>
      <frequency>daily</frequency>
      <qemu>
        <xscreen>800x600x24</xscreen>
        <binary>qemu</binary>
        <options>-cdrom</options>
        <pause>10</pause>
        <sendkeys>kp_enter</sendkeys>
        <!-- Time to run qmeu -->
        <time>600</time>
      </qemu>
      <output>
        <!-- Number of test sets to keep -->
        <keep>4</keep>
        <root>/home/bms/autotesting/tests/</root>
        <local>example-live/squeeze/xfce/</local>
        <video>autotesting.ogv</video>
        <screenshots>
          <final>final-screenshot.png</final>
          <montage>montage-of-video-frames.png</montage>
        </screenshots>
      </output>
    </test>
  </tests>
</autotesting>


LOOKING AT EACH TAG
===================

<download>http://127.0.0.1/autotesting/debian-live-501-i386-xfce-desktop.iso</download>
The location of the image to be tested. This can be an iso or an USB image (img) or in fact anything that can be booted with qemu.

<title>Debian Live (xfce)</title>
<description>Debian Live. Daily build of squeeze xfce iso.</description>
The title and description tags are used in the opening titles of the video and also useful when scanning a long tests.xml file. 

<background>http://git.debian.org/?p=debian-live/homepage.git;a=blob_plain;f=images/debian-live.png</background>
The location of the background graphic used in the video. 

<frequency>daily</frequency>
This can be daily, weekly or monthly. Daily, means that the test will be run every time. Weekly - the test will only be done if the current day is a Sunday. Monthly - the test will only be done on the 1st day of a month, 

<xscreen>800x600x24</xscreen>
The width, height and colour depth of the xserver (Xvfb) in which to run qemu. Not at the minimum dimensions required for qemu are 800x600. 

<binary>qemu</binary>
The qemu binary. This could be for example qemu-ppc64 or qemu-arm, to test images for different architectures. 

<options>-cdrom</options>
The option to pass to qemu in conjunction with the image we are testing. In this example we are testing a iso, so we need the "-cdrom" option. For an img file you would use -hda.

<pause>10</pause>
Time to pause before sending keys presses.

<sendkeys>kp_enter</sendkeys>
Some iso and img require some user interaction to star the booting, normally just a enter being pressed. Useful for selecting different options from the grub menu. Use the qmeu sendkey command, send "kp_enter" - a keypad enter. To send multiple key presses, use a comma separated list. <sendkeys>down,down,kp_enter</sendkeys>. 

<time>600</time>
Let qemu run for 600 seconds.

<keep>4</keep>
Number of test sets to keep. If frequency is weekly, and keep is 4, then any test set older than 4 weeks will be deleted.

<root>/home/bms/autotesting/tests/</root>
<local>example-live/squeeze/xfce/</local>
A new directory will be created at /root/prefix/DATE/, in this example /home/bms/autotesting/tests/example-live/squeeze/xfce/2009-07-01/ and the output files will be stored in this directory. Another directory /root/current/prefix/ will be created, with symlinks to the latest output files, e.g. /home/bms/autotesting/tests/current/example-live/squeeze/xfce/.

<video>autotesting.ogv</video>
The output video file stored in the output directories (see <root> and <local>). File is a theora video, created by the wonderful application - recordmydeskptop.

<final>final-screenshot.png</final>
<montage>montage-of-video-frames.png</montage>
The output file names stored in the output directories (see <root> and <local>) of a final screenshot of qemu and a montage of frames from the video. These can give a good idea of how successful the test was rather than viewing the whole video. 


Some example test may be found in /usr/share/autotesting/tests/


debian/changelog
autotesting (0.0.1-1.2) unstable; urgency=low

  * Adding 60alpha1 tests
  * Adding md5sum cheak

 -- Brendan M. Sleight <bms@brum>  Fri, 26 Mar 2010 22:01:55 +0000

autotesting (0.0.1-1.1) unstable; urgency=low

  * Adding some useful test templates 

 -- Brendan M. Sleight <bms@barwap.com>  Sun, 05 Jul 2009 23:40:36 +0100

autotesting (0.0.1-1) unstable; urgency=low

  * Initial release 

 -- Brendan M. Sleight <bms@barwap.com>  Wed, 01 Jul 2009 21:03:14 +0100



autotesting.sh
#!/bin/sh
#
# Script to be run video-qemu-boot-iso.sh via cron job
#

while getopts p:v:q:t: opt
do
    case "$opt" in
      p)  VQBI="$OPTARG";;
      q)  QEMU_BIN="$OPTARG";;
      t)  TIME_Q="$OPTARG";;
      v)  VQUALITY="$OPTARG";;


      \?)		# unknown flag
      	  echo >&2 \
		"usage: $0 [-p path to video-qemu-booting-iso.sh ] [-q alternative qemu binary name] [-t time to run qemu] /foo/bar/iso.directory/ /foo/bar/video.directory/ "
	  exit 1;;
    esac
done
shift `expr $OPTIND - 1`

if [ -z "$1" ] && [ -z "$2" ] && [ -z "$3" ]; then
    cat << EOF
usage: $0 [-p path to video-qemu-booting-iso.sh ] [-q alternative qemu binary name] [-t time to run qemu] [-v (0 to 10) encoding quality for video] /foo/bar/iso.directory/ /foo/bar/video.directory/ 

 This script runs the video-qemu-booting-iso.sh for each iso in a directory. 

 This script will take _along_ time to run. Approx 1 hour per iso.
 Basically for time to run = number of isos x time to run qmeu x 3  seconds
 For testing the set-up use -t with a small value e.g. -t 10

 The use of -q could allow qemu-ppc be used once Debian bug #388735 is resolved.
EOF
    exit 0
fi

ISODIR=$1
VIDEODIR=$2

VQBI="${VQBI:-video-qemu-booting-iso.sh}"
QEMU_BIN="${QEMU:-qemu}" #qemu_0.8.4-etch1
TIME_Q="${TIME_Q:-1200}"
VQUALITY="${VQUALITY:-5}"

for ISO in $ISODIR/*.iso
do
 ISOBN=$(basename $ISO)
 VIDEO=$VIDEODIR$ISOBN.ogg
 echo "Running $VQBI -t $TIME_Q -q $QEMU_BIN -v $VQUALITY $ISO $VIDEO"
 $VQBI -t $TIME_Q -q $QEMU_BIN -v $VQUALITY $ISO $VIDEO
done
echo "Finished Autotesting"


clean-old-files.sh
#!/bin/sh
#
# Remove old files
find /home/autotesting/*/daily/ -not -type d -ctime +4 -print0     | xargs --null --no-run-if-empty rm -f
find /var/www/debian-live/i386/*-sid-*  -not -type d -ctime +4 -print0   | xargs --null --no-run-if-empty rm -f
find /var/www/webconverger/i386/*  -not -type d -ctime +4 -print0   | xargs --null --no-run-if-empty rm -f
find /var/www/local-build/debian-live/i386/*  -not -type d -ctime +4 -print0   | xargs --null --no-run-if-empty rm -f
find /var/www/lenny_live_beta1/i386/*  -not -type d -ctime +4 -print0   | xargs --null --no-run-if-empty rm -f

find /home/autotesting/*/weekly/ -not -type d -ctime +28 -print0   | xargs --null --no-run-if-empty rm -f
find /var/www/debian-live/i386/*-lenny-*  -not -type d -ctime +28 -print0   | xargs --null --no-run-if-empty rm -f

find /home/autotesting/*/monthly/ -not -type d -ctime +124 -print0 | xargs --null --no-run-if-empty rm -f
find /var/www/debian-live/i386/*-etch-*  -not -type d -ctime +4 -print0   | xargs --null --no-run-if-empty rm -f

# Remove empty directories
find /var/www/*/ -depth -type d -empty -exec rmdir {} \;


daily-cron-script.sh
#!/bin/sh
#
# Neater then one long cron line
#

BASEDIR="/home/autotesting"

cd ${BASEDIR}/debian-live/autotesting

exec >${BASEDIR}/log/daily/autotesting.log 2>&1
echo "New Day $(date)"

# lenny_live_beta1.lst
./run-batch-autotesting.sh lenny_live_beta1.lst ${BASEDIR}/iso/daily/ ${BASEDIR}/video/daily/ /var/www/lenny_live_beta1/i386/

# Test "daily" images
./run-batch-autotesting.sh daily.lst ${BASEDIR}/iso/daily/ ${BASEDIR}/video/daily/ /var/www/debian-live/i386/

# Test webcoverger
./run-batch-autotesting.sh webconverger.lst ${BASEDIR}/iso/daily/ ${BASEDIR}/video/daily/ /var/www/webconverger/i386/ webc-3.2.mini.iso.MD5SUM

# Build our own images
exec >${BASEDIR}/log/daily/autobuild.log 2>&1
for DIST in etch lenny sid
do
    ./local_build_debian-live.sh ${BASEDIR}/iso/local/debian-live ${DIST} xfce
done

# Test our own images
exec >>${BASEDIR}/log/daily/autotesting.log 2>&1
./run-batch-autotesting.sh local.lst ${BASEDIR}/iso/local/ ${BASEDIR}/video/local/ /var/www/local-build/debian-live/i386/


run-batch-autotesting.sh
#!/bin/sh
#
# run-batch-autotesting.sh                                #
#                                                         #
# Copyleft (c) 2008 Brendan M. Sleight                    #
#              <bmsleight@barwap._REMOVE_SPAM_TRAP.com>   #
#                                                         #
# This script is licensed under the GNU GPL v2 or later.  #
#                                                         #
# On Debian systems, this license can be obtained via     #
# /usr/share/common-licenses/GPL                          #
#
# Download the list of images and run video-qemu-booting against each download.
#
# Usage: run-batch-autotesting.sh filelist.txt isolocation videos
#   e.g: run-batch-autotesting.sh dailylist.txt daily/ video/daily/
# Hence can be run Daily, weekly or month using different lists
# Another script can do the tidy up.
#
#   We make be downloading an iso on the 21st Jan, built on the
# 20th or even the 1st Jan. This needs to be identified in the
# someway - mayeb the filename. 
# iso-file-name_Built_02-Apr-2008_tested_03-Apr-2008.ogg
#
# 0 - Check lockfile
# 1 - For each iso [each file in in Auotest-iso-urls.lst]
# 1a -  Remove the old iso
# 1b -  if file does not exist or datestamp local iso iso is before today:
# 1b -    Download the new-current-iso and related md5sum-file
# 1c -  check md5sum, if md5sum ok:
# 1d -    Video-qemu-booting-iso iso-name date-iso-name.ogg/.jpg
# 1e -    ????
# 2  -  Profit
#
#

VIDEO_QEMU_BOOTING="/home/autotesting/debian-live/autotesting/video-qemu-booting-iso.sh"

if [ -z "$1" ] && [ -z "$2" ] && [ -z "$3" ] && [ -z "$4" ]
then
    echo "usage: $0 directory/ "
    echo
    echo " e.g: run-batch-autotesting.sh dailylist.txt iso/daily/ video/daily/ debian-live/i386/ MD5SUM-FILE"
    exit
fi

URLLIST=$1
DIRECTORY=$2
VIDEO_DIRECTORY=$3
WEB_ROOT=$4

if [ -z "$5" ]
then
    MD5SUM_FILE="MD5SUMS"
else
    MD5SUM_FILE="$5"
fi

if [ -f /tmp/run-batch-autotesting.lock ]
then
   echo "Lock file /tmp/run-batch-autotesting.lock present indicating $0 is already running"
   exit
fi

echo "Lock file present indicating $0 is running" >/tmp/run-batch-autotesting.lock

for URL in $(cat $URLLIST)
do 
    BASE_NAME=$(basename $URL)
    PART_URL=$(dirname $URL)
    MD5SUMS="$PART_URL/$MD5SUM_FILE"
    HAS_FILE_UPDATED_TODAY=$(find $DIRECTORY/ -ctime -0 \! -type d | grep "$BASE_NAME")

    if [ -n "$HAS_FILE_UPDATED_TODAY" ]
    then
        echo "$BASE_NAME already downloaded today "
        echo " - skipping download and testing. To retest please remove file or set ctime>1 day"
        continue
    fi

    case "${URL}" in
        http*)
            echo "Downloading $URL"
            rm $DIRECTORY/$BASE_NAME 2>/dev/null
            rm $DIRECTORY/$MD5SUM_FILE 2>/dev/null
            wget --no-verbose --tries=3 --timeout=60 --directory-prefix=$DIRECTORY $URL
            wget --no-verbose --tries=3 --timeout=60 --directory-prefix=$DIRECTORY $MD5SUMS
            MD5SUM_LOCAL=$(md5sum $DIRECTORY/$BASE_NAME | cut --fields=1 --delimiter=\  )
            MD5SUM_REMOTE=$(cat $DIRECTORY/$MD5SUM_FILE | grep "$BASE_NAME" | head -n 1 | cut --fields=1 --delimiter=\  )
            ;;
        *)
            echo "Local file - assuming already in place"
            MD5SUM_LOCAL="ok"
            MD5SUM_REMOTE="$MD5SUM_LOCAL"
            ;;
    esac

    if [ $MD5SUM_LOCAL != $MD5SUM_REMOTE ]
    then
        echo "$BASE_NAME - md5sums different, remote: $MD5SUM_REMOTE, local: $MD5SUM_LOCAL. "
        echo "    Skip testing."
        continue
    fi

    DATE_IMAGE_BUILT=$(ls -lh --time-style long-iso $DIRECTORY/$BASE_NAME |tr -s " "|cut -d" " -f6)
    DATE_DOWNLOADED=$(ls -lc --time-style long-iso $DIRECTORY/$BASE_NAME |tr -s " "|cut -d" " -f6)
    TODAY=$(date  +"%F")
    TODAY_LONG=$(date )
    mkdir "$VIDEO_DIRECTORY/${BASE_NAME}" 
    VIDEO_NAME="$VIDEO_DIRECTORY/${BASE_NAME}/Built_${DATE_IMAGE_BUILT}_Tested_${TODAY}_.ogg"
    LOG_FILE="$VIDEO_DIRECTORY/${BASE_NAME}/Built_${DATE_IMAGE_BUILT}_Tested_${TODAY}_log.txt"
    echo "AutoTesting $BASE_NAME $VIDEO_NAME"
    $VIDEO_QEMU_BOOTING -g 1280x960  -t 420 -v 5  $DIRECTORY/$BASE_NAME $VIDEO_NAME >$LOG_FILE 2>&1
    echo "Finished Autotesting $BASE_NAME"
    echo
    echo "----"
    echo
    echo "* Moving to webserver"
    LN_S_ROOT="$WEB_ROOT/$BASE_NAME/built_${DATE_IMAGE_BUILT}/Tested_${TODAY}"
    LN_S_CURRENT="$WEB_ROOT/$BASE_NAME/current"
    # making directories if required.
    mkdir -p "$LN_S_ROOT"
    rm $LN_S_CURRENT -rf
    mkdir -p "$LN_S_CURRENT"
    # get rid of the old current
    cp $VIDEO_NAME                "$LN_S_ROOT/video-booting.theora_${TODAY}.ogg"
    mv $VIDEO_NAME             "$LN_S_CURRENT/video-booting.theora_${TODAY}.ogg"
    cp $VIDEO_NAME.montage.jpg    "$LN_S_ROOT/video-booting-montage_${TODAY}.jpg"
    mv $VIDEO_NAME.montage.jpg "$LN_S_CURRENT/video-booting-montage_${TODAY}.jpg"
    cp $VIDEO_NAME.end.jpg        "$LN_S_ROOT/final-screenshot_${TODAY}.jpg"
    mv $VIDEO_NAME.end.jpg     "$LN_S_CURRENT/final-screenshot_${TODAY}.jpg"
    echo "----"
    cp /home/autotesting/debian-live/autotesting/homepage/video-page.wml "$LN_S_ROOT/index.wml"
    cp /home/autotesting/debian-live/autotesting/homepage/video-page.wml "$LN_S_CURRENT/index.wml"
    DETAILS_TXT="$VIDEO_DIRECTORY/${BASE_NAME}/Built_${DATE_IMAGE_BUILT}_Tested_${TODAY}__details.txt"
    echo "# Setting used for index.wml"   >$DETAILS_TXT
    echo "url=${URL}"                     >>$DETAILS_TXT
    echo "image=${BASE_NAME}"             >>$DETAILS_TXT
    echo "built=${DATE_IMAGE_BUILT}"      >>$DETAILS_TXT
    echo "tested=${TODAY}"                >>$DETAILS_TXT
    echo "tested_long=\"${TODAY_LONG}\" " >>$DETAILS_TXT
    cp $DETAILS_TXT              "$LN_S_ROOT/details.txt"
    mv $DETAILS_TXT              "$LN_S_CURRENT/details.txt"
    echo
done



rm /tmp/run-batch-autotesting.lock


autotesting.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
#   autotesting - automatically test by video qemu booting.
#   Copyright (C) Brendan M. Sleight, et al. <bms@barwap.com>
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.


from optparse import OptionParser
import amara, datetime, os, shutil, subprocess, sys, telnetlib, tempfile, time, datetime

def displayNumber():
    """Really a global we can change later to something better.
    like automaticaly get the next free display number. """
    return ":8"
    
def log(message):
    """ Print [date] message """ 
    now = datetime.datetime.now()
    print "[" + str(now) + "] " + str(message) 
    sys.stdout.flush()

def cronCheak(frequency, debug):
    """Check if the frequency (e.g. daily, weekly, mounthly is due to 
    be run at this time
    returns boolean.
    """
    now = datetime.datetime.now()
    if frequency == "Daily" or frequency == "daily":
        return True
    elif frequency == "Weekly" or frequency == "weekly":
        if now.weekday() == 0:
            return True
    elif frequency == "Monthly" or frequency == "monthly":
        if now.day == 1:
            return True
    elif debug == True:
        log("*** Debug mode - run regardless of frequency ***")
        return True
    return False

def dateBasedOnFrequency(frequency, keep):
    now = datetime.datetime.now()
    if frequency == "Daily" or frequency == "daily":
        days=1 
    elif frequency == "Weekly" or frequency == "weekly":
        days=7 
    elif frequency == "Monthly" or frequency == "monthly":
        # No date delta of month so near enough ...
        days=30
    old = now - datetime.timedelta(days=days*keep)
    return old

def parseTest(xml):
    """ Get the xml file and load as a object usign amara
    returns amaraObject"""
    tests = amara.parse(xml)
    return tests.autotesting.tests

def wget(url, limit):
    """ Use wget to download a file
    return fileObject"""
    # --limit-rate=4M
    log("Downloading " + url)
    tmpFile = tempfile.NamedTemporaryFile(prefix="autotesting_wget_")
    L = ['wget', '-nv', str(url), '-O', tmpFile.name]
    if limit == True:
        L.append('--limit-rate=2M')
    # Maybe do something with the retcode in the future.
    retcode = subprocess.call(L)
    return tmpFile

def getDownloads(download, backgroundURL, limit):
    """ Create two temp files and download the main download and 
    the background 
    return tempfile, tempfile""" 
    background = wget(str(backgroundURL), limit)
    try:
        # datetime.timedelta(days=55) 
        daysOfset = int(str(download.daysofset)) 
        dateUrlPart = datetime.datetime.now() - datetime.timedelta(days=daysOfset)
        downloadURL=str(download.url) + dateUrlPart.strftime(str(download.dateformat)) + str(download.dateformatend)
    except:
        downloadURL=str(download.url)
    download = wget(str(downloadURL), limit)
    return download, downloadURL, background

def testCheckSum(download, downloadURL, md5sum):
    """ Check md5sum locally against remote list of md5sum stored. 
    """
    mdFile = wget(str(md5sum), False)
    data = [line.split() for  line in open(mdFile.name,'r')]
    downloadFilename = downloadURL.split('/')[-1]
    required = False
    for line in data:
        if line[1] == downloadFilename:
            log("MDSUM required: " + str(line))
            required = line
    if required:
        p = subprocess.Popen("md5sum " + download.name, shell = True, stdout=subprocess.PIPE) 
        localMD = p.stdout.readline().split()
        p.wait()
        log("MDSUM required: " + str(localMD))
        if localMD[0] == required[0]:
            log("MDSUM matched")
            return True
        else:
            log("MDSUM not matched")
            return False
    else:
        log("MDSUM not found for filename")
        return False
            
def authorityFile():
    """ Return a file containing "localhost"
    """
    authority = tempfile.NamedTemporaryFile(prefix="autotesting_authority_")
    authority.write("localhost\n")
    authority.close()
    return authority

def startXvfb(display, xscreen, background):
    """Start a Xvfb instance. The host display for qemu to run inside.
    then set the background image. Return ths PID of Xvfb
    return interger"""
    log("Starting Xvfb")
    authority = authorityFile()
    xvfbCommand = ["Xvfb", display, "-auth", authority.name, "-screen", "0", xscreen] 
    xvfb = subprocess.Popen(xvfbCommand)
    time.sleep(1)
    log("Setting Xvfb background")
    xloadimage = ["xloadimage", "-display", display, "-onroot", "-fullscreen", background]
    retcode = subprocess.call(xloadimage)
    return xvfb

def waitUntilEncodingFinished(recordMyDesktop):
    log("Waiting for recordmydesktop encoding to finish")
    while recordMyDesktop.poll()!=0:
        print ".. ",
        time.sleep(1)
    print " "

def kill(process, message):
    log(message)
    kill = ["kill", str(process.pid)]
    retcode = subprocess.call(kill)

def startRecordMyDesktop(display):
    log("Starting Record My Desktop")
    video = tempfile.NamedTemporaryFile(prefix="autotesting_video_", suffix=".ogv")
    recordmydesktopCommand = ["recordmydesktop",  "--no-cursor", "-display", 
                               display ,"--no-sound", "--overwrite", "-o" ,video.name]
    recordmydesktop = subprocess.Popen(recordmydesktopCommand)
    return recordmydesktop, video 

def xmessage(display, title, message, time, font):
    xloadimage = ["gmessage", "-display", display, "-timeout", time, 
                "-font", font, "-button", "", "-title", title, 
                "-center", "\n" + message]
    retcode = subprocess.call(xloadimage)
    
def openingTitles(display, test, downloadURL):
    log("Showing Opening Titles")
    time.sleep(2)
    xmessage(display, str(test.title), "Autotesting of: " + str(test.title), "3", "monospace 14")
    xmessage(display, str(test.title), "Created at " + str(datetime.datetime.now()), "2", "monospace 10")
    xmessage(display, str(test.title), str(test.description), "3", "monospace 6")
    xmessage(display, str(test.title), "Downloaded using this url: " + str(downloadURL), "2", "monospace 6")
    
def runningQemu(display, test, qemuDownload):
    """ Start qemu running, with a local telnet port at 55555 listening acting as the qemu monitor
    returns subprocess.process"""
    telnet = ("127.0.0.1", "55555")
    address, port = telnet
    monitor = "telnet:" + address + ":" + port  + ",server,nowait"
    qemuBinary = str(test.qemu.binary)
    log("Starting " + qemuBinary)
    if not qemuBinary.startswith("qemu"):
        log("WARNING! : " + qemuBinary + " does not start with qemu, using qemu instead.")
        qemuBinary = "qemu"
    qemuCommand = [qemuBinary, "-monitor", monitor, "-full-screen", str(test.qemu.options), str(qemuDownload)]
    qemu = subprocess.Popen(qemuCommand, env={"DISPLAY": display})
    sendkeysToQemu(test, telnet)
    log("Running qemu for  " + str(test.qemu.time) + " seconds")
    time.sleep(int(str(test.qemu.time)))
    return qemu

def sendkeysToQemu(test, telnet):
    """Using telnet to send commands to qemu monitor
    """
    address, port = telnet
    log("Sending keys after pause of " + str(test.qemu.pause) + " seconds")
    time.sleep(int(str(test.qemu.pause)))
    tn = telnetlib.Telnet(address, port)
    sendkeys = str(test.qemu.sendkeys)
    for key in sendkeys.split(','):
        log("qemu sendkey " + key)
        tn.write("sendkey " + key + "\n")
        time.sleep(1)


def captureScreenshot(display):
    """ Capture screen shot on display
    return fileObject"""
    log("Capture screenshot")
    finalImage = tempfile.NamedTemporaryFile(prefix="autotesting_video_", suffix=".png")
    captureCommand = ["import", "-display", display, "-window", "root", finalImage.name]
    retcode = subprocess.call(captureCommand)
    return finalImage

def createMontage(video, test):
    """ Make 16 frames of the video at 1/16, 2/16, 3/16 .... 16/16 of 
        way through the video. The make a montage of these frames in to
        one image. 
    return fileObject"""
    log("Creating Montage")
    montage = tempfile.NamedTemporaryFile(prefix="autotesting_video_", suffix=".png")
    videoLength = int(str(test.qemu.pause)) + int(str(test.qemu.time)) 
    # Time for opening titles
    videoLength = videoLength + 2 + 3 + 3 + 2
    listFrames = []
    listFramesNames = []
    for count in range(16):
        ss = 0.0625 * count * videoLength
        frame = tempfile.NamedTemporaryFile(prefix="autotesting_frame_", suffix=".jpg")
        frameName = frame.name + "%d.jpg"
        frameNameOut = frame.name + "1.jpg"
        ffmpeg = ["ffmpeg", "-i", video.name, "-an", "-ss", str(ss), "-t", 
                  "01", "-r", "1", "-y", frameName]
        retcode = subprocess.call(ffmpeg)
        try:
            shutil.move(frameNameOut, frame.name)
            listFramesNames.append(frame.name)
            listFrames.append(frame)
        except:
            pass
    montageCommand = ["montage", "-geometry", "180x135+4+4", "-frame", "5"]
    for frame in listFramesNames:
        montageCommand.append(frame)
    montageCommand.append(montage.name)
    retcode = subprocess.call(montageCommand)
    return montage

def storeFile(tmpFile, copyLocation, symLocation):
    shutil.copyfile(tmpFile, copyLocation)
    try:
        os.remove(symLocation)
    except:
        pass
    os.symlink(copyLocation, symLocation)

def fileOutputs(test, video, finalImage, montage):
    """ Move files to the correct place (as per XML tags)
    including symlinks to current.
    """
    log("Moving files to the correct place.")
    root=str(test.output.root)
    local=str(test.output.local)
    today=str(datetime.date.today())
    mainDir = root + "/" + local + "/" + today + "/"
    dateDirs = root + "/" + local + "/"
    currentDir = root + "/current/" + local + "/"
    try:
        os.makedirs(mainDir)
    except:
        pass
    try:
        os.makedirs(currentDir)
    except:
        pass
    storeFile(video.name, mainDir + str(test.output.video), 
                                   currentDir + str(test.output.video))    
    storeFile(finalImage.name, mainDir + str(test.output.screenshots.final), 
                                   currentDir + str(test.output.screenshots.final))
    storeFile(montage.name, mainDir + str(test.output.screenshots.montage), 
                                   currentDir + str(test.output.screenshots.montage))
    keepDate = dateBasedOnFrequency(str(test.frequency), int(str(test.output.keep)))
    log("Removing old Autotesting output before " + keepDate.strftime("%A %B %d %I:%M:%S %p %Y"))
    for f in os.listdir(dateDirs):
        fp = dateDirs + f
        if (time.mktime(keepDate.timetuple()) - os.path.getmtime(fp) ) > 0:
            # Remove the old file
            shutil.rmtree(fp)
            log("Removing - " + fp)

def main():
    usage = "usage: %prog [options] --tests=TESTS.XML \n       %prog --help for all options"
    parser = OptionParser(usage, version="%prog ")
    parser.add_option("-t", "--tests", dest="tests",
                      help="complete the autotesting definined in the xml template")
    parser.add_option("-d", "--debug", action="store_true", dest="debug",
                      help="Debug, run all tests regardless of frquency of testing.")
    parser.add_option("-l", "--limit", action="store_true", dest="limit",
                      help="Limit download rates to 4M")
    (options, args) = parser.parse_args() 
    if not options.tests :
        parser.error("Must pass a list of tests to complete.")
        
    # Main loop
    display = displayNumber()
    tests = parseTest(options.tests)
    for test in tests.test:
        if cronCheak(str(test.frequency), options.debug):
            log("Starting Autotesting of: " + str(test.title))
            (download, downloadURL, background) = getDownloads(test.download, str(test.background), options.limit)
            try:
                md5sum = str(test.download.md5sum)
                checkSum = testCheckSum(download, downloadURL, md5sum)
            except:
                log("No md5sum found - assuming file downloaded ok")
                checkSum = True
            if checkSum == True:
                xvfb = startXvfb(display, str(test.qemu.xscreen), background.name)
                (recordMyDesktop, video) = startRecordMyDesktop(display)
                openingTitles(display, test, downloadURL)
                try:
                    qemu = runningQemu(display, test, download.name)
                    finalImage = captureScreenshot(display)
                    kill(qemu, "Killing qemu")
                except:
                    xmessage(display, str(test.title), "Qemu failed to run correctly!", "5", "monospace 14")
                    finalImage = captureScreenshot(display)
                    log("Qemu failed")
                kill(recordMyDesktop, "Killing recordmysdesktop")
                waitUntilEncodingFinished(recordMyDesktop)
                kill(xvfb, "Killing Xvfb")
                # Put video, montage in right place remvoe old versions etc.
                montage = createMontage(video, test)
                fileOutputs(test, video, finalImage, montage)
            else:
                log("Checksum (md5sum) failed - test skipped")
            log("Finished Autotesting of: " + str(test.title))
            log("*****************************")
        else:
            log("Skipping Autotesting of: " + str(test.title) + " as frequency " + str(test.frequency))
            log("*****************************")

if __name__ == "__main__":
    main()
